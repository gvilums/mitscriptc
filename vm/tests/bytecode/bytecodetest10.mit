x = 1;

f = fun(){
	x = {
		x : 1;
	};
	this = {
		increment1 : fun(y){
			g = fun(){
				x = x + y;
				global x;
				return x;
			};
			return g;
		};
		increment2 : fun(y){
			g = fun(){
				x.x = x.x + y;
				return x.x;
			};
			return g;
		};
		get_val : fun(){
			return x.x;
		};
	};
	return this;
};

fact = f();

f1 = fact.increment1(1);
f2 = fact.increment1(2);
f3 = fact.increment1(3);

print(x + " " + fact.get_val());
print(f1() + " " + fact.get_val());
print(f2() + " " + fact.get_val());
print(f3() + " " + fact.get_val());

f1 = fact.increment2(1);
f2 = fact.increment2(2);
f3 = fact.increment2(3);

print(x + " " + fact.get_val());
print(x + " " + f1());
print(x + " " + f2());
print(x + " " + f3());

