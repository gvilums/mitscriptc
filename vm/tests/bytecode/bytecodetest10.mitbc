function
{
	functions =
	[
		function
		{
			functions = [],
			constants = [],
			parameter_count = 1,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [],
			instructions = 
			[
			]
		},
		function
		{
			functions = [],
			constants = [],
			parameter_count = 0,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [],
			instructions = 
			[
			]
		},
		function
		{
			functions = [],
			constants = [],
			parameter_count = 1,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [],
			instructions = 
			[
			]
		},
		function
		{
			functions =
			[
				function
				{
					functions =
					[
						function
						{
							functions = [],
							constants = [None],
							parameter_count = 0,
							local_vars = [],
							local_ref_vars = [],
							free_vars = [y],
							names = [x],
							instructions = 
							[
								load_global	0
								push_ref	0
								load_ref
								add
								store_global	0
								load_global	0
								return
								load_const	0
								return
							]
						}
					],
					constants = [None],
					parameter_count = 1,
					local_vars = [y, g],
					local_ref_vars = [y],
					free_vars = [],
					names = [],
					instructions = 
					[
						load_func	0
						push_ref	0
						alloc_closure	1
						store_local	1
						load_local	1
						return
						load_const	0
						return
					]
				},
				function
				{
					functions =
					[
						function
						{
							functions = [],
							constants = [None],
							parameter_count = 0,
							local_vars = [],
							local_ref_vars = [],
							free_vars = [x, y],
							names = [x, x, x],
							instructions = 
							[
								push_ref	0
								load_ref
								push_ref	0
								load_ref
								field_load	0
								push_ref	1
								load_ref
								add
								field_store	1
								push_ref	0
								load_ref
								field_load	2
								return
								load_const	0
								return
							]
						}
					],
					constants = [None],
					parameter_count = 1,
					local_vars = [y, g],
					local_ref_vars = [y],
					free_vars = [x],
					names = [],
					instructions = 
					[
						load_func	0
						push_ref	1
						push_ref	0
						alloc_closure	2
						store_local	1
						load_local	1
						return
						load_const	0
						return
					]
				},
				function
				{
					functions = [],
					constants = [None],
					parameter_count = 0,
					local_vars = [],
					local_ref_vars = [],
					free_vars = [x],
					names = [x],
					instructions = 
					[
						push_ref	0
						load_ref
						field_load	0
						return
						load_const	0
						return
					]
				}
			],
			constants = [1, None],
			parameter_count = 0,
			local_vars = [x, this],
			local_ref_vars = [x],
			free_vars = [],
			names = [x, increment1, increment2, get_val],
			instructions = 
			[
				push_ref	0
				alloc_record
				dup
				load_const	0
				field_store	0
				store_ref
				alloc_record
				dup
				load_func	0
				alloc_closure	0
				field_store	1
				dup
				load_func	1
				push_ref	0
				alloc_closure	1
				field_store	2
				dup
				load_func	2
				push_ref	0
				alloc_closure	1
				field_store	3
				store_local	1
				load_local	1
				return
				load_const	1
				return
			]
		}
	],
	constants = [1, 1, 2, 3, " ", " ", " ", " ", 1, 2, 3, " ", " ", " ", " "],
	parameter_count = 0,
	local_vars = [],
	local_ref_vars = [],
	free_vars = [],
	names = [print, input, intcast, x, f, fact, f1, increment1, f2, increment1, f3, increment1, get_val, get_val, get_val, get_val, increment2, increment2, increment2, get_val],
	instructions = 
	[
		load_func	0
		alloc_closure	0
		store_global	0
		load_func	1
		alloc_closure	0
		store_global	1
		load_func	2
		alloc_closure	0
		store_global	2
		load_const	0
		store_global	3
		load_func	3
		alloc_closure	0
		store_global	4
		load_global	4
		call	0
		store_global	5
		load_global	5
		field_load	7
		load_const	1
		call	1
		store_global	6
		load_global	5
		field_load	9
		load_const	2
		call	1
		store_global	8
		load_global	5
		field_load	11
		load_const	3
		call	1
		store_global	10
		load_global	0
		load_global	3
		load_const	4
		add
		load_global	5
		field_load	12
		call	0
		add
		call	1
		load_global	0
		load_global	6
		call	0
		load_const	5
		add
		load_global	5
		field_load	13
		call	0
		add
		call	1
		load_global	0
		load_global	8
		call	0
		load_const	6
		add
		load_global	5
		field_load	14
		call	0
		add
		call	1
		load_global	0
		load_global	10
		call	0
		load_const	7
		add
		load_global	5
		field_load	15
		call	0
		add
		call	1
		load_global	5
		field_load	16
		load_const	8
		call	1
		store_global	6
		load_global	5
		field_load	17
		load_const	9
		call	1
		store_global	8
		load_global	5
		field_load	18
		load_const	10
		call	1
		store_global	10
		load_global	0
		load_global	3
		load_const	11
		add
		load_global	5
		field_load	19
		call	0
		add
		call	1
		load_global	0
		load_global	3
		load_const	12
		add
		load_global	6
		call	0
		add
		call	1
		load_global	0
		load_global	3
		load_const	13
		add
		load_global	8
		call	0
		add
		call	1
		load_global	0
		load_global	3
		load_const	14
		add
		load_global	10
		call	0
		add
		call	1
	]
}