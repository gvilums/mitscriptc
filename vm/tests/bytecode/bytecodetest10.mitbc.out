1 1
2 1
4 1
7 1
7 1
7 2
7 4
7 7
